components/style/style_resolver.rs

```diff
@@ -298,6 +298,28 @@ where
         })
     }
 
+    /// Cascade a set of rules for pseudo element, using the default parent for inheritance.
+    pub(crate) fn cascade_style_and_visited_for_pseudo_with_default_parents(
+        &mut self,
+        inputs: CascadeInputs,
+        pseudo: &PseudoElement,
+        primary_style: &PrimaryStyle,
+    ) -> ResolvedStyle {
+        with_default_parent_styles(self.element, |_, layout_parent_style| {
+            let layout_parent_style_for_pseudo = if primary_style.style().is_display_contents() {
```

Comment:
Can we reuse the logic here somehow? I guess it's not too much of an issue, but I think I'd prefer a `fn layout_parent_style_for_pseudo(&PrimaryStyle, Option<&ComputedStyle>)` or what not

Reply:
For sure. I've added a new helper function. Let me know if it's what you were thinking.